File: apps/ios/Sources/SrsStore.swift
import Foundation

struct SrsState: Codable {
    let intervalDays: Int
    let due: Date
}

struct WrongEntry: Codable, Hashable {
    let dict: String
    let type: String
    let practice: String
}

struct WrongToday: Codable {
    let date: String
    let items: [WrongEntry]
}

final class SrsStore {
    static let shared = SrsStore()

    private let srsKey = "jlpt-n4-srs"
    private let wrongKey = "jlpt-n4-wrong-today"
    private let statsKey = "jlpt-n4-stats"

    func loadSrs() -> [String: SrsState] {
        guard let data = UserDefaults.standard.data(forKey: srsKey) else { return [:] }
        return (try? JSONDecoder().decode([String: SrsState].self, from: data)) ?? [:]
    }

    func saveSrs(_ value: [String: SrsState]) {
        if let data = try? JSONEncoder().encode(value) {
            UserDefaults.standard.set(data, forKey: srsKey)
        }
    }

    func loadWrongToday() -> WrongToday {
        let today = Stats.todayKey()
        guard let data = UserDefaults.standard.data(forKey: wrongKey),
              let stored = try? JSONDecoder().decode(WrongToday.self, from: data)
        else {
            return WrongToday(date: today, items: [])
        }
        if stored.date != today {
            return WrongToday(date: today, items: [])
        }
        return stored
    }

    func saveWrongToday(_ value: WrongToday) {
        if let data = try? JSONEncoder().encode(value) {
            UserDefaults.standard.set(data, forKey: wrongKey)
        }
    }

    func loadStats() -> Stats {
        guard let data = UserDefaults.standard.data(forKey: statsKey),
              let stats = try? JSONDecoder().decode(Stats.self, from: data)
        else {
            return Stats()
        }
        return stats
    }

    func saveStats(_ value: Stats) {
        if let data = try? JSONEncoder().encode(value) {
            UserDefaults.standard.set(data, forKey: statsKey)
        }
    }
}

struct Stats: Codable {
    var streak: Int = 0
    var todayCount: Int = 0
    var lastDate: String = Stats.todayKey()

    static func todayKey() -> String {
        let formatter = DateFormatter()
        formatter.dateFormat = "yyyy-MM-dd"
        return formatter.string(from: Date())
    }

    mutating func normalizeForToday() {
        let today = Stats.todayKey()
        if lastDate != today {
            lastDate = today
            todayCount = 0
        }
    }
}

File: apps/android/app/src/main/java/com/learnjapanese/app/data/SrsStore.kt
package com.learnjapanese.app.data

import android.content.Context
import kotlinx.serialization.builtins.MapSerializer
import kotlinx.serialization.builtins.serializer

class SrsStore(
    context: Context,
) {
    private val prefs = context.getSharedPreferences("learnjapanese.srs", Context.MODE_PRIVATE)
    private val json = FixtureLoader.json()

    private val srsKey = "jlpt-n4-srs"
    private val wrongKey = "jlpt-n4-wrong-today"
    private val statsKey = "jlpt-n4-stats"

    fun loadSrs(): Map<String, SrsState> {
        val raw = prefs.getString(srsKey, null) ?: return emptyMap()
        val serializer = MapSerializer(String.serializer(), SrsState.serializer())
        return runCatching { json.decodeFromString(serializer, raw) }.getOrElse { emptyMap() }
    }

    fun saveSrs(value: Map<String, SrsState>) {
        val serializer = MapSerializer(String.serializer(), SrsState.serializer())
        val raw = json.encodeToString(serializer, value)
        prefs.edit().putString(srsKey, raw).apply()
    }

    fun loadWrongToday(): WrongToday {
        val today = Stats.todayKey()
        val raw = prefs.getString(wrongKey, null) ?: return WrongToday(date = today, items = emptyList())
        val stored =
            runCatching { json.decodeFromString(WrongToday.serializer(), raw) }.getOrNull()
                ?: return WrongToday(date = today, items = emptyList())
        return if (stored.date != today) {
            WrongToday(date = today, items = emptyList())
        } else {
            stored
        }
    }

    fun saveWrongToday(value: WrongToday) {
        val raw = json.encodeToString(WrongToday.serializer(), value)
        prefs.edit().putString(wrongKey, raw).apply()
    }

    fun loadStats(): Stats {
        val raw = prefs.getString(statsKey, null) ?: return Stats()
        return runCatching { json.decodeFromString(Stats.serializer(), raw) }.getOrElse { Stats() }
    }

    fun saveStats(value: Stats) {
        val raw = json.encodeToString(Stats.serializer(), value)
        prefs.edit().putString(statsKey, raw).apply()
    }
}
